Postgres and Azure Blob Storage backend docs (#29688)
gravitational/teleport: docs/pages/reference/backends.mdx

     etcd_max_client_msg_size_bytes: 15728640
```

## PostgreSQL (Preview)

<Admonition type="warning" title="Preview">

>   The PostgreSQL cluster state and audit log storage is currently in Preview.

Reviewer: Should we also note that it's only available in 13.3+?

Author: Added a version warning in cdc9b599e7ed9646c8139890fdf884cd4cd7ed59

</Admonition>

<Admonition
  type="tip"
  title="Tip"
>
  Before continuing, please make sure to take a look at the [Cluster State section](../architecture/nodes.mdx#cluster-state)
  in Teleport Architecture documentation.
</Admonition>

Teleport can use [PostgreSQL](https://www.postgresql.org/) as a storage back-end
to achieve high availability. You must take steps to protect access to
PostgreSQL in this configuration because that is where Teleport secrets like
keys and user records will be stored. The PostgreSQL backend supports two types
of Teleport data:

- Cluster state
- Audit log events

The PostgreSQL backend requires PostgreSQL 13 or later, and, for the cluster

> state only, the [`wal2json`](https://github.com/eulerto/wal2json) logical

Reviewer 2: Can we add information/example on how to install this plugin and check whether it's installed?

Author: 86b1e7a6ed8d62b09e9ca9c231d7120e539d1027 adds installation references, unfortunately plugin loading is just `dlopen`, so there is no way to list available plugins without actually attempting to load them.

decoding plugin. Teleport needs separate databases for the cluster state and the
audit log, and it will attempt to create them if given permissions to do so; it

> will also set up the database schemas as needed, so

Reviewer: incomplete sentence?

Author: missed a change in the commit, ed4374c405

The PostgreSQL backend for cluster state relies on the ability to use [logical
decoding](https://www.postgresql.org/docs/current/logicaldecoding.html) to get a
stream of changes from the database; because of that, the
[`wal_level`](https://www.postgresql.org/docs/current/runtime-config-wal.html#GUC-WAL-LEVEL)
parameter must be set to `logical` and

> [`max_replication_slots`](https://www.postgresql.org/docs/current/runtime-config-replication.html#GUC-MAX-REPLICATION-SLOTS)

Reviewer 2: Can we give example config snippet with these parameter?

must be set to at least as many Teleport auth servers as you'll be running (a
higher number is recommended, to account for network timeouts and
reconnections). In addition, the database user must have the `initiating
replication` [role
attribute](https://www.postgresql.org/docs/current/role-attributes.html) (as

> well as `login privilege`, of course). Since replication permissions allow for

Reviewer 2: Can we give example SQL command to create user with necessary permissions, or update an existing one?

    # pool_max_conns is an additional parameter that determines the maximum
    # number of connections in the connection pool used for the cluster state
    # database (the change feed uses an additional connection), defaulting to
    # a value that depends on the number of available CPUs.
    conn_string: postgresql://user_name@database-address/teleport_backend?sslmode=verify-full&pool_max_conns=20

    # An audit_events_uri with a scheme of postgresql:// will use the
    # PostgreSQL backend for audit log storage; the URI is a libpq-compatible
    # connection string just like the cluster state conn_string, but cannot be
    # specified as key=value pairs. It's possible to specify completely
    # different PostgreSQL clusters for cluster state and audit log.
    audit_events_uri:
      - postgresql://user_name@database-address/teleport_audit?sslmode=verify-full
```

Audit log events are periodically deleted after a default retention period of
3766 hours (one year); it's possible to select a different retention period or
to disable the cleanup entirely, by specifying the `retention_period` or the

> `disable_cleanup` parameters in the fragment of the URI:

Reviewer: Why are these in the fragment and not just additional query params?

Author: I didn't want to accidentally interfere with some other parameter that one might want to set in the connection string itself; this way we have a distinction between Teleport parameters and database connection parameters.

```yaml
storage:
  teleport:
    audit_events_uri:
      - postgresql://user_name@database-address/teleport_audit?sslmode=verify-full#disable_cleanup=false&retention_period=2160h
```

### Authentication

We strongly recommend using [client
certificates](https://www.postgresql.org/docs/current/ssl-tcp.html#SSL-CLIENT-CERTIFICATES)
to authenticate Teleport to PostgreSQL, as well as enforcing the use of TLS and
[verifying the server
certificate](https://www.postgresql.org/docs/current/libpq-ssl.html#LIBQ-SSL-CERTIFICATES)
on the client side. If the use of passwords is unavoidable, we recommend
configuring them in [the `~/.pgpass`
file](https://www.postgresql.org/docs/current/libpq-pgpass.html) rather than

> storing them in the configuration file.

Reviewer: ```suggestion
storing them in Teleport's configuration file.
```

Reviewer 2: If we provide an option to keep passwords in Teleport config file but not recommending to use it, can we just remove that option and always require .pgpass or support reading them from files in Teleport config?

Author: That option comes from `pgx`'s config string parsing, do we really want to start mucking around with that?

Reviewer 2: Ah right, for some reason I thought we have a separate fileconfig option in Teleport to provide these. If we don't and it's just a connection string parsing option, then it's fine.

Author: I considered checking that the password is blank in the resulting configuration, but the .pgpass file and the PGPASSWORD envvar are checked at parsing time too, so we can only prohibit passwords altogether.

### Azure AD authentication

If you are running Teleport on Azure, Teleport can make use of [Azure AD
authentication](https://learn.microsoft.com/azure/postgresql/flexible-server/concepts-azure-ad-authentication)
to connect to an [Azure Database for
PostgreSQL](https://azure.microsoft.com/en-us/products/postgresql) server
without having to manage any secrets:

```yaml
teleport:
  storage:
    type: postgresql

    conn_string: postgresql://user_name@database-name.postgres.database.azure.com/teleport_backend?sslmode=verify-full&pool_max_conns=20

>     auth_mode: azure

Reviewer 2: Can we default to azure mode for Azure databases? Or even just mandate Azure IAM for them?

Author: In one of the RFDs we agreed that we don't want to special-case the `.postgres.database.azure.com` suffix in case they change it in the future (maybe for some variation of their cloud?).

The chart will have a dedicated `azure` configuration mode, FWIW.

     etcd_max_client_msg_size_bytes: 15728640
```

## PostgreSQL (Preview)

<Details
  title="Version warning"

>   opened={true}

Author: @ptgott should we default close this in the v14+ docs and leave it open by default on the v13 docs?

will also set up the database schemas as needed, so we recommend giving the user
ownership over the databases.

The PostgreSQL backend for cluster state relies on the ability to use [logical
decoding](https://www.postgresql.org/docs/current/logicaldecoding.html) to get a
stream of changes from the database; because of that, the
[`wal_level`](https://www.postgresql.org/docs/current/runtime-config-wal.html#GUC-WAL-LEVEL)
parameter must be set to `logical` and
[`max_replication_slots`](https://www.postgresql.org/docs/current/runtime-config-replication.html#GUC-MAX-REPLICATION-SLOTS)
must be set to at least as many Teleport auth servers as you'll be running (a
higher number is recommended, to account for network conditions). `wal_level`
can only be set at server start, so it should be set in `postgresql.conf`:
```
# the default value for wal_level is replica
wal_level = logical

# the default value for max_replication_slots is 10
max_replication_slots = 10
```

In addition, the database user must have the `initiating replication` [role
attribute](https://www.postgresql.org/docs/current/role-attributes.html) (as

> well as `login privilege`, of course). In the `psql` shell:

Author: Removed in 34f0aac8d8bf991337fada08c039ffbe0a1bfc65

```
postgres=# CREATE USER new_user WITH REPLICATION;
CREATE ROLE

postgres=# ALTER ROLE existing_user WITH LOGIN REPLICATION;
ALTER ROLE
```

Since replication permissions allow for essentially full read access over the
entire cluster (with a physical replication connection) or to all databases that
the user can connect to, it's recommended to prevent the user from opening
replication connections, or from connecting to databases other than the ones
used for Teleport, if the PostgreSQL cluster is shared between Teleport and
other applications.

For convenience, Teleport will attempt to grant itself the `initiating

> replication` role attribute, to accomodate the ability of some managed services

Author: ```suggestion
replication` role attribute, to accommodate the ability of some managed services
```

status` user, or similar privileged access as can be granted in managed
PostgreSQL services such as Azure Databases for PostgreSQL or Amazon RDS for

> PostgreSQL).

Reviewer 3: The parenthetical phrase is longer than the rest of the paragraph—I think we can restructure this to make it clearer. How does Teleport test whether it can grant itself the `initiating replication` role attribute?

Author: Simplified while also actually describing the purpose behind it in 0172239d3145327b25466a1a2946f6dcec123636, PTAL.

To configure Teleport to use PostgreSQL:

> - Configure all Teleport Auth servers to use the PostgreSQL back-end in the

Reviewer 3: ```suggestion
- Configure all Teleport Auth Service instances to use the PostgreSQL back-end in the
```

  `storage` section of `teleport.yaml` as shown below.

> - Deploy several auth servers connected to the PostgreSQL storage back-end.

Reviewer 3: ```suggestion
- Deploy several Auth Service instances connected to the PostgreSQL storage back-end.
```

> - Deploy several proxy nodes.

Reviewer 3: ```suggestion
- Deploy several Proxy Service nodes.
```

> - Make sure that the proxy instances and all Teleport resource services that

Reviewer 3: ```suggestion
- Make sure that the Proxy Service instances and all Teleport agent services that
```

>   connect directly to the the auth service have the `auth_server` configuration

Reviewer 3: ```suggestion
  connect directly to the the Auth Service have the `auth_server` configuration
```

>   setting populated with the address of a load balancer for auth service

Reviewer 3: ```suggestion
  setting populated with the address of a load balancer for Auth Service
```
